src/solc_0.8/asset/libraries/AssetHelper.sol

```diff
@@ -0,0 +1,96 @@
+//SPDX-License-Identifier: MIT
+pragma solidity 0.8.2;
+
+import "../../catalyst/interfaces/IAssetAttributesRegistry.sol";
+
+// used to reduce PolygonAssetV2 contract code size
+library AssetHelper {
+    struct AssetRegistryData {
+        IAssetAttributesRegistry assetRegistry;
+    }
+
+    function setCatalystDatas(
+        AssetRegistryData storage self,
+        IAssetAttributesRegistry.AssetGemsCatalystData[] memory assetGemsCatalystData
+    ) public {
+        for (uint256 i = 0; i < assetGemsCatalystData.length; i++) {
+            require(assetGemsCatalystData[i].catalystContractId > 0, "WRONG_catalystContractId");
+            require(assetGemsCatalystData[i].assetId != 0, "WRONG_assetId");
+            self.assetRegistry.setCatalystWhenDepositOnOtherLayer(
+                assetGemsCatalystData[i].assetId,
+                assetGemsCatalystData[i].catalystContractId,
+                assetGemsCatalystData[i].gemContractIds
+            );
+        }
+    }
+
+    function decodeAndSetCatalystDataL1toL2(AssetRegistryData storage self, bytes calldata depositData)
+        public
+        returns (
+            uint256[] memory ids,
+            uint256[] memory amounts,
+            bytes32[] memory hashes
+        )
+    {
+        bytes memory data;
+        IAssetAttributesRegistry.AssetGemsCatalystData[] memory catalystDatas;
+        (ids, amounts, data) = abi.decode(depositData, (uint256[], uint256[], bytes));
+        (hashes, catalystDatas) = abi.decode(data, (bytes32[], IAssetAttributesRegistry.AssetGemsCatalystData[]));
+
+        setCatalystDatas(self, catalystDatas);
+    }
+
+    function decodeAndSetCatalystDataL2toL1(AssetRegistryData storage self, bytes calldata data)
```

Comment:
When going to L2 There's the ids and amount encoded in extra data going back there's no amount and ids because it's called by the predicate and it has already these infos.